torch/fx/graph.py

```diff
@@ -252,6 +252,288 @@ class _PyTreeInfo(NamedTuple):
     in_spec: pytree.TreeSpec
     out_spec: Optional[pytree.TreeSpec]
 
+@compatibility(is_backward_compatible=False)
+class CodeGen(object):
+    def __init__(self):
+        self._body_transformer: Optional[TransformCodeFunc] = None
+
+    def generate_prologue(self, free_vars: List[str], maybe_return_annotation: str) -> str:
+        """
+        Given the free variables and a return annotation, generates the beginning of the FX function.
+        By default, `generate_prologue(['a', 'b'], '') == 'def forward(a, b):'`
+        """
+        # If the original function didn't have self as its first argument, we
+        # would have added it.
+        if len(free_vars) == 0 or free_vars[0] != 'self':
+            free_vars.insert(0, 'self')
+        return f"def forward({', '.join(free_vars)}){maybe_return_annotation}:"
+
+    def generate_output(self, output_args: Argument) -> str:
+        """
+        Given the output arguments, generates the return statement of the FX function.
+        """
+        return f'return {repr(output_args)}'
+
+    def process_inputs(self, *args: Any) -> Any:
+        """
+        Transforms the inputs so that the graph can take them as arguments, as
+        non-default codegen may result in the inputs to the function being
+        different from the inputs to the graph.
+
+        If the graph was directly runnable, this invariant should hold true
+        `f.process_outputs(f.graph(*f.process_inputs(*inputs))) == f(*inputs)`
+        """
+        return args
+
+    def process_outputs(self, outputs: Any) -> Any:
+        """
+        Transforms the outputs of the graph to be identical to the codegen.
+
+        See ``process_inputs`` for more details.
+        """
+        return outputs
```

Comment:
If I understand correctly, `process_{in,out}puts` are called within a runtime context rather than a codegen context. Thus, that these are inside a class named `CodeGen` seems kind-of strange (and also taking a `self` argument for a `CodeGen` instance is kinda strange). Also, the example on line `284` accesses `process_inputs` and `process_outputs` on `f`, but I think you would actually have to access it on `f.graph.process_inputs` (assuming `f` is a `GraphModule`). We might want to think of some organization where the runtime aspects and the codegen aspects are put in different places. e.g. `CodeGen` could have `register_input_transformer(t : Callable[[Any], ...])` where it takes in the function to apply to the inputs, then during the codegen process that function is assigned to an attribute like `GraphModule.input_transformer` that is dispatched in `GraphModule.__call__`. Similarly for `output_transformer`. Then I think the runtime aspects live on the GraphModule, which is the thing that gives a Graph its runtime semantics, rather than living on the graph itself

Reply:
The reason they live here is that when you override the `CodeGen` class, this is also what you need to override in order to get it to process your inputs to work "properly" with your graph (for example, I think currently fx.Transformer doesn't work with the pytree stuff).

This isn't actually needed for the GraphModule runtime *except* if you're trying to process your GraphModule inputs to give you your inputs to the graph.

Reply:
Ideally, perhaps, the codegen functionality wouldn't be on the graph at all, but that might be a slightly larger refactor.